File: server/game/deck.go
package game

import "math/rand"

// Deck holds an ordered collection of cards.
type Deck struct {
	Cards []Card
}

// NewDeck creates a 112-card deck.
//
// Per color (Red, Yellow, Green, Blue):
//   - Numbers 1–9 ×2 (no zero)
//   - Skip ×2, Reverse ×2, DrawTwo ×2
//   - Swap ×1 (colored)
//
// Global (wild):
//   - Wild ×4, WildDrawFour ×4, GlobalSwitch ×4
//
// Per color = 2*9 + 2 + 2 + 2 + 1 = 25; 4 colors = 100; +12 wild = 112.
func NewDeck() *Deck {
	cards := make([]Card, 0, 112)

	for _, col := range []Color{Red, Yellow, Green, Blue} {
		for v := 1; v <= 9; v++ {
			cards = append(cards, Card{Color: col, Kind: Number, Value: v})
			cards = append(cards, Card{Color: col, Kind: Number, Value: v})
		}
		for i := 0; i < 2; i++ {
			cards = append(cards, Card{Color: col, Kind: Skip})
			cards = append(cards, Card{Color: col, Kind: Reverse})
			cards = append(cards, Card{Color: col, Kind: DrawTwo})
		}
		cards = append(cards, Card{Color: col, Kind: Swap})
	}

	for i := 0; i < 4; i++ {
		cards = append(cards, Card{Color: Wild, Kind: WildCard})
		cards = append(cards, Card{Color: Wild, Kind: WildDrawFour})
		cards = append(cards, Card{Color: Wild, Kind: GlobalSwitch})
	}

	return &Deck{Cards: cards}
}

// Shuffle randomizes the deck in place using the provided rng. A nil rng falls
// back to the global package source, which is for tests and nothing else: every
// shuffle that decides a hand takes the room's crypto-seeded source, Replenish
// included.
func (d *Deck) Shuffle(rng *rand.Rand) {
	swap := func(i, j int) { d.Cards[i], d.Cards[j] = d.Cards[j], d.Cards[i] }
	if rng != nil {
		rng.Shuffle(len(d.Cards), swap)
		return
	}
	rand.Shuffle(len(d.Cards), swap)
}

// Draw removes and returns the top card of the deck.
func (d *Deck) Draw() (Card, bool) {
	if len(d.Cards) == 0 {
		return Card{}, false
	}
	top := d.Cards[len(d.Cards)-1]
	d.Cards = d.Cards[:len(d.Cards)-1]
	return top, true
}

// DrawN removes and returns n cards from the top of the deck, or nothing at all
// when fewer than n are left. Dealing is the one caller that wants this: a short
// hand is a broken round, so it is better to know before any card moves.
//
// Every in-play draw wants DrawUpTo instead.
func (d *Deck) DrawN(n int) ([]Card, bool) {
	if len(d.Cards) < n {
		return nil, false
	}
	drawn := make([]Card, n)
	for i := 0; i < n; i++ {
		drawn[i], _ = d.Draw()
	}
	return drawn, true
}

// DrawUpTo removes and returns however many of the n requested cards are left,
// possibly none.
//
// A draw in this game can never fail. Once every card sits in a hand there is
// nothing to reshuffle and nothing to hand over, and the only sane answer is a
// smaller penalty: an error at this point would have to be returned to a player
// who has no other legal action, which freezes the round for the whole table.
// The caller decides what "nothing came out" means for the turn.
func (d *Deck) DrawUpTo(n int) []Card {
	if n > len(d.Cards) {
		n = len(d.Cards)
	}
	if n <= 0 {
		return nil
	}
	drawn := make([]Card, n)
	for i := 0; i < n; i++ {
		drawn[i], _ = d.Draw()
	}
	return drawn
}

// Replenish replaces the deck with the shuffled discard pile. The caller keeps
// the current top card out of `discard` — it stays on the pile so the round
// still has something to match against.
//
// The rng is the room's, like every other shuffle that decides what lands in a
// hand. It used to be the global source, on the argument that only the deal was
// reconstructible; but the pile going back into the deck is the second half of
// a long round, every card in it has been seen by the table, and an attacker who
// can predict its order knows the rest of the round outright. A shuffle nobody
// may predict has no business being convenient.
func (d *Deck) Replenish(discard []Card, rng *rand.Rand) {
	d.Cards = make([]Card, len(discard))
	copy(d.Cards, discard)
	d.Shuffle(rng)
}

File: client/src/content/deck.ts
/**
 * The deck, as the tables and the catalogue a content page needs: what is in it,
 * what each card does, and what it costs you to still be holding it.
 *
 * One structure rather than three because the composition, the effects and the
 * scoring are the same eight kinds seen from three angles, and pages that listed
 * them separately would eventually disagree with each other. The numbers come
 * from `docs/rules.md` §2 and §10, and `contentPages.test.ts` checks them
 * against `server/game/deck.go` and `server/game/card.go`, which are the only
 * authorities in the repo.
 *
 * The *names* deliberately do not come from `docs/rules.md`: that file is the
 * implementation spec and still calls these by their SOLO names ("Miss a Turn",
 * "Change Cards All Round"). The player-facing names live in `t.cardNames`.
 *
 * Build-time only, like everything under `src/content/`: this never ships to a
 * player, so it does not grow the bundle that does.
 */
import type { CardKind } from '../types/protocol'
import type { Lang } from '../seo/meta'

export interface DeckEntry {
  kind: CardKind
  /** How many of this kind are in the 112. */
  copies: number
  /**
   * Points it is worth in somebody else's hand when the round ends. `null` for
   * numbers, which are worth their face value and so have no single number.
   */
  points: number | null
  /** How the copies break down, in each language. */
  detail: Record<Lang, string>
  /**
   * What the card does, as a catalogue entry.
   *
   * Deliberately its own wording rather than the matching line of `t.rules`:
   * that line is a rulebook bullet read in sequence ("Skip: the next player
   * loses their turn."), this one is read on its own by somebody who came
   * looking for one card. Mapping a kind to a bullet would mean parsing a name
   * out of a sentence, which breaks on the first rewording. The *facts* are not
   * duplicated: copies and points are checked against the server.
   */
  effect: Record<Lang, string>
}

export const DECK: readonly DeckEntry[] = [
  {
    kind: 'number',
    copies: 72,
    points: null,
    detail: {
      en: '4 colours × values 1 to 9 × 2 copies',
      fr: '4 couleurs × valeurs 1 à 9 × 2 exemplaires',
    },
    effect: {
      en: 'Plays on a card of the same colour or the same number, and does nothing else. Most of a hand is these, and most of a turn is finding one that fits.',
      fr: 'Se pose sur une carte de la même couleur ou du même chiffre, et ne fait rien de plus. L’essentiel d’une main, et l’essentiel d’un tour.',
    },
  },
  {
    kind: 'skip',
    copies: 8,
    points: 20,
    detail: { en: '4 colours × 2 copies', fr: '4 couleurs × 2 exemplaires' },
    effect: {
      en: 'The next player loses their turn. In a duel that means the turn comes straight back to you.',
      fr: 'Le joueur suivant saute son tour. En duel, la main te revient donc aussitôt.',
    },
  },
  {
    kind: 'reverse',
    copies: 8,
    points: 10,
    detail: { en: '4 colours × 2 copies', fr: '4 couleurs × 2 exemplaires' },
    effect: {
      en: 'Play changes direction. With two players there is nobody to go around, so it behaves like a second Skip.',
      fr: 'Le jeu repart dans l’autre sens. À deux, il n’y a personne à contourner : ça revient à un Passe.',
    },
  },
  {
    kind: 'draw_two',
    copies: 8,
    points: 30,
    detail: { en: '4 colours × 2 copies', fr: '4 couleurs × 2 exemplaires' },
    effect: {
      en: 'The next player draws two, unless they answer with a +2 of their own and pass the whole stack along. Taking the cards does not cost them the turn: they draw, then play or pass.',
      fr: 'Le suivant pioche deux cartes, sauf s’il répond par un +2 et fait suivre toute la pile. Piocher ne lui coûte pas son tour : il pioche, puis joue ou passe.',
    },
  },
  {
    kind: 'swap',
    copies: 4,
    points: 30,
    detail: { en: '4 colours × 1 copy', fr: '4 couleurs × 1 exemplaire' },
    effect: {
      en: 'Played on your turn like any coloured card. Pick anyone at the table and take their whole hand; they get yours.',
      fr: 'Se joue à ton tour comme n’importe quelle carte colorée. Tu désignes quelqu’un et tu prends toute sa main ; il récupère la tienne.',
    },
  },
  {
    kind: 'global_switch',
    copies: 4,
    points: 40,
    detail: { en: 'no colour: it plays on anything', fr: 'sans couleur : elle se pose sur tout' },
    effect: {
      en: 'Name the new colour, then every hand at the table slides one seat along. Nobody keeps anything.',
      fr: 'Tu annonces la nouvelle couleur, puis chaque main glisse d’une place. Personne ne garde rien.',
    },
  },
  {
    kind: 'wild',
    copies: 4,
    points: 40,
    detail: { en: 'no colour: it plays on anything', fr: 'sans couleur : elle se pose sur tout' },
    effect: {
      en: 'Lands on anything, whatever colour is on the pile. You name the colour that follows.',
      fr: 'Se pose sur n’importe quoi, quelle que soit la couleur en cours. C’est toi qui annonces la couleur suivante.',
    },
  },
  {
    kind: 'wild_draw_four',
    copies: 4,
    points: 50,
    detail: { en: 'no colour: it plays on anything', fr: 'sans couleur : elle se pose sur tout' },
    effect: {
      en: 'Lands on anything, names the colour, and the next player draws four unless they stack a +4 of their own.',
      fr: 'Se pose sur n’importe quoi, annonce la couleur, et le suivant pioche quatre cartes s’il ne cumule pas un +4.',
    },
  },
]

/** 112. Asserted in a test, because the tables are only right if this is. */
export const DECK_SIZE = DECK.reduce((n, e) => n + e.copies, 0)

/** The four suits, in the order the card art fans them. */
export const SUITS = ['red', 'yellow', 'green', 'blue'] as const

/** Entry by kind, for a page that walks the deck in its own order. */
export function entry(kind: CardKind): DeckEntry {
  const found = DECK.find((e) => e.kind === kind)
  if (!found) throw new Error(`no deck entry for ${kind}`)
  return found
}
